pkg/cnsmigration/migrator.go

```diff
@@ -0,0 +1,315 @@
+	matchingCnsVolumes []types.CnsVolume
+	usedVolumeCache    *inUseVolumeStore
+
+	// for counting migrated volumes and stuff
+	migratedVolumes int
+	volumesNotFound int
+	failedToMigrate int
+}
+
+func NewCNSVolumeMigrator(config *rest.Config, dsSource, dsTarget string) *CNSVolumeMigrator {
+	return &CNSVolumeMigrator{
+		kubeConfig:           config,
+		sourceDatastore:      dsSource,
+		destinationDatastore: dsTarget,
+	}
+}
+
+func (c *CNSVolumeMigrator) createOpenshiftClients() {
+	// create the clientset
+	c.clientSet = kubernetes.NewForConfigOrDie(c.kubeConfig)
+
+	c.openshiftConfigClientSet = configclient.NewForConfigOrDie(rest.AddUserAgent(c.kubeConfig, clientOperatorName))
+	c.openshiftOperatorClientSet = operatorclient.NewForConfigOrDie(rest.AddUserAgent(c.kubeConfig, clientOperatorName))
+}
+
+func (c *CNSVolumeMigrator) StartMigration(ctx context.Context, volumeFile string) error {
+	// find existing CSI based persistent volumes, which same datastore as one mentioned in the source
+	c.createOpenshiftClients()
+
+	err := c.loginToVCenter(ctx)
+	if err != nil {
+		printError("error logging into vcenter: %v", err)
+		return err
+	}
+
+	printGreenInfo("logging successfully to vcenter")
+
+	err = c.getCNSVolumes(ctx, c.sourceDatastore)
+	if err != nil {
+		printError("error listing cns volumes: %v", err)
+		return err
+	}
+	pvList, err := c.clientSet.CoreV1().PersistentVolumes().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	podList, err := c.clientSet.CoreV1().Pods("").List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	c.usedVolumeCache = NewInUseStore(pvList.Items)
+	c.usedVolumeCache.addAllPods(podList.Items)
+	err = c.findAndMigrateCSIVolumes(ctx, volumeFile)
+	c.printSummary()
+	return err
+}
+
+func (c *CNSVolumeMigrator) printSummary() {
+	printGreenInfo("----------- Migration Summary ------------")
+	printGreenInfo("Migrated %d volumes", c.migratedVolumes)
+	printGreenInfo("Failed to migrate %d volumes", c.failedToMigrate)
+	printGreenInfo("Volumes not found %d", c.volumesNotFound)
+}
+
+func (c *CNSVolumeMigrator) findAndMigrateCSIVolumes(ctx context.Context, volumeFile string) error {
+	file, err := os.Open(volumeFile)
+	if err != nil {
+		msg := fmt.Errorf("error opening file %s: %v", volumeFile, err)
+		klog.Error(msg)
+		return msg
+	}
+	defer file.Close()
+
+	fileBytes, err := io.ReadAll(file)
+	if err != nil {
+		msg := fmt.Errorf("error reading file %s: %v", volumeFile, err)
+		klog.Error(msg)
+		return msg
+	}
+	if len(fileBytes) == 0 {
+		msg := fmt.Errorf("file %s has no listed volumes", volumeFile)
+		return msg
+	}
+
+	volumeLines := strings.Split(string(fileBytes), "\n")
+	for _, volumeLine := range volumeLines {
+		pvName := strings.TrimSpace(volumeLine)
+		if pvName == "" {
+			continue
+		}
+		printGreenInfo("Starting migration for pv %s", pvName)
+		pv, err := c.clientSet.CoreV1().PersistentVolumes().Get(ctx, pvName, metav1.GetOptions{})
```

Comment:
This can be very big. Our supported maximum is 150 000 Pods. I tested with ~130 000 pods (listing all pods on a small cluster 500 times and concatenating the result), the migrator took 5GB of RSS memory.

Only a fraction of Pods will use PVs. Would it make sense to sort PVs by PVC namespace and then go by namespace by namespace?

At least a TODO item would be nice.